### ISSUE TITLE
深度解读 Observation —— SwiftUI 性能提升的新途径 | 肘子的Swift记事本

### ISSUE THREAD

**Issue by fatbobman** [fatbobman/blogComments#195]
https://www.fatbobman.com/posts/mastering-Observation/ 

 在 WWDC 2023 中，苹果介绍了 Swift 标准库中的新成员：Observation 框架。它的出现有望缓解开发者长期面临的 SwiftUI 视图无效更新问题。本文将采取问答的方式，全面而详尽地探讨 Observation 框架，内容涉及其产生原因、使用方法、工作原理以及注意事项等。

**Comment by skywalkerlw** [comment]
在“测试一：” 中，为啥store.a = 100不会引起onChange？

**Comment by fatbobman** [comment]
> 在“测试一：” 中，为啥store.a = 100不会引起onChange？

因为在执行 store.b = 100 时，便已经调用了 onChange 闭包。该闭包只会调用一次

**Comment by moekyo** [comment]
Macro应该是只能增加代码，而不能对源代码做修改或者删除的吧？[Macros 说明](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/macros/#:~:text=Expanding%20a%20macro%20is%20always%20an%20additive%20operation%3A%20Macros%20add%20new%20code%2C%20but%20they%20never%20delete%20or%20modify%20existing%20code.)

**Comment by fatbobman** [comment]
> Macro应该是只能增加代码，而不能对源代码做修改或者删除的吧？[Macros 说明](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/macros/#:~:text=Expanding%20a%20macro%20is%20always%20an%20additive%20operation%3A%20Macros%20add%20new%20code%2C%20but%20they%20never%20delete%20or%20modify%20existing%20code.)

是的，这个部分出现了错误。谢谢

**Comment by mcmay** [comment]
When SwiftData comes into the scene to play, things will change. SwiftData seems to add the observation by default via the @Model macro so the @Observation macro becomes superfluous.

**Comment by fatbobman** [comment]
> When SwiftData comes into the scene to play, things will change. SwiftData seems to add the observation by default via the @model macro so the @observation macro becomes superfluous.

Yes, `@Model` covers the functionalities implemented by the `@Observable` macro. In the article [Unveiling the Data Modeling Principles of SwiftData](https://fatbobman.com/en/posts/unveiling-the-data-modeling-principles-of-swiftdata/), I described the specific implementation details.

**Comment by mcmay** [comment]
> @fatbobman
> > When SwiftData comes into the scene to play, things will change. SwiftData seems to add the observation by default via the @model macro so the @observation macro becomes superfluous.
> 
> Yes, `@Model` covers the functionalities implemented by the `@Observable` macro. In the article [Unveiling the Data Modeling Principles of SwiftData](https://fatbobman.com/en/posts/unveiling-the-data-modeling-principles-of-swiftdata/), I described the specific implementation details.

Thanks for sharing. I've read that article before this one. It's an in-depth treatment of the modelling mechanism of SwiftData. By the way, the question about the runtime exception of my app project raised after that article has now been resolved. It turned out to be the misplacement of a .sheet() modifier in a wrong place in my code, although I'm still at a loss why the exception happened because that. There is surely a lot to be explored in the interplay between SwiftUI and SwiftData.

**Comment by Shinolr** [comment]
> 比如，当前开发者通常会使用结构体（ Struct ）来构建应用的状态模型。使用了 Observation 框架后，为了实现属性级别的观察，我们应该改用 Observation 框架创建可观察对象，甚至多层嵌套的对可观察对象来构建状态模型。

苹果一直[推荐](https://developer.apple.com/documentation/swift/choosing-between-structures-and-classes)使用`struct`作为model的类型。但是越来越多的官方sample code使用`class`构建model，最近几年尝试wwdc推出的新东西时，总是不知道该用`struct`还是`class`。看完文章，终于解开疑惑，知道如何选型了，谢肘大。

**Comment by guoyingtao** [comment]
发现一个typo `@Publsiehd`
```
@StateObject var store = Store() // 只要有属性（ @Publsiehd ）发生变化，便对 body 重新评估
```

非常棒的文章 👍🏻 多谢分享！

**Comment by fatbobman** [comment]
> 发现一个typo `@Publsiehd`

@guoyingtao  谢谢，已修改

**Comment by opserx** [comment]
我觉得有种场景可能很常见。一个模型我需要使用SwiftData存储，同时我也需要也他是@Obsevable。这种情况怎么做？需要定义两个类么？

**Comment by fatbobman** [comment]
@opserx 没明白。SwiftData 的 Model 本身就是使用了 Observation ，如果你的意思是，希望其中的部分属性基于 Observation 框架，但是不想持久化，那么就用 @Transient 来标注它。尽管这样不是最推荐的，但也基本上够用了。

**Comment by opserx** [comment]
谢谢。是我糊涂，没想道SwiftData 的 Model 本身就是使用了 Observation。这么简单居然不知道。

**Comment by Lightning-Lion** [comment]
关于withObservationTracking，它可以观察诞生自@Observable的class的@Binding值。比如这个例子：
```swift
import SwiftUI

@MainActor
@Observable
class ViewModel {
    var userID = UUID()
}

struct ContentView: View {
    @State
    private var vm = ViewModel()
    var body: some View {
        VStack {
            SubView0(userID: $vm.userID)
            Button("New User", action: {
                vm.userID = UUID()
            })
        }
        .padding()
        .task(id: "onLoad", {
            let _ = withObservationTracking({
                vm.userID
            }, onChange: {
                print("[ContentView]:User Id Change")
            })
        })
    }
}

struct SubView0: View {
    @Binding
    var userID:UUID
    var body: some View {
        Text("SubView0")
        .task(id: "onLoad", {
            let _ = withObservationTracking({
                userID
            }, onChange: {
                print("[SubView0]:User Id Change")
            })
        })
    }
}
```

**Comment by Lightning-Lion** [comment]
关于withObservationTracking，你可以使用这种递归的语法来创建持续响应值变化的行为，和onChange一样，但避免了onChange的[这个Bug](https://fatbobman.com/en/posts/the-anomaly-of-onchange-in-swiftui-multi-layer-navigation/)
```swift
import SwiftUI

@MainActor
@Observable
class ViewModel {
    var userID = UUID()
}

struct ContentView: View {
    @State
    private var vm = ViewModel()
    var body: some View {
        VStack {
            SubView0(userID: $vm.userID)
            Button("New User", action: {
                vm.userID = UUID()
            })
        }
        .padding()
    }
}

struct SubView0: View {
    @Binding
    var userID:UUID
    var body: some View {
        Text("SubView0")
        .task(id: "onLoad", {
            onChangeForValue0(perform: {
                print("[SubView0]:User Id Change")
            })
        })
    }
    func onChangeForValue0(perform:@escaping ()->()) {
        let _ = withObservationTracking({
            userID
        }, onChange: {
            perform()
            // withObservationTracking不保证onChange调用的线程，因此我们总是回到主线程
            Task { @MainActor in
                onChangeForValue0(perform: perform)
            }
        })
    }
}
```

**Comment by gongzhang** [comment]
请问使用 let 和 var 声明的属性（不仅仅是这次讲的可观察对象）是否有细微区别？

**Comment by fatbobman** [comment]
@gongzhang 在swiftui 的视图中，没有区别。

**Comment by gongzhang** [comment]
> [@gongzhang](https://github.com/gongzhang) 在swiftui 的视图中，没有区别。

感谢回复☺️
